packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts

```diff
@@ -0,0 +1,221 @@
+import type { TSESLint, TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+
+import * as util from '../util';
+
+const nodeTypesAreEqual = <T extends TSESTree.Node>(
+  actualTypeNode: TSESTree.Node,
+  expectedTypeNode: T,
+): actualTypeNode is T => {
+  return actualTypeNode.type === expectedTypeNode.type;
+};
+
+const constituentsAreEqual = (
+  actualTypeNode: TSESTree.TypeNode,
+  expectedTypeNode: TSESTree.TypeNode,
+): boolean => {
+  switch (expectedTypeNode.type) {
+    // These types should never occur as constituents of a union/intersection
+    case AST_NODE_TYPES.TSAbstractKeyword:
+    case AST_NODE_TYPES.TSAsyncKeyword:
+    case AST_NODE_TYPES.TSDeclareKeyword:
+    case AST_NODE_TYPES.TSExportKeyword:
+    case AST_NODE_TYPES.TSNamedTupleMember:
+    case AST_NODE_TYPES.TSOptionalType:
+    case AST_NODE_TYPES.TSPrivateKeyword:
+    case AST_NODE_TYPES.TSProtectedKeyword:
+    case AST_NODE_TYPES.TSPublicKeyword:
+    case AST_NODE_TYPES.TSReadonlyKeyword:
+    case AST_NODE_TYPES.TSRestType:
+    case AST_NODE_TYPES.TSStaticKeyword:
+    case AST_NODE_TYPES.TSTypePredicate:
+      /* istanbul ignore next */
+      throw new Error(`Unexpected Type ${expectedTypeNode.type}`);
+
+    default:
+      if (nodeTypesAreEqual(actualTypeNode, expectedTypeNode)) {
+        return astNodesAreEquals(actualTypeNode, expectedTypeNode);
+      }
+      return false;
```

Comment:
Nit: expressing the intent of the code a little bit more... what do you think?

```suggestion
      return (
        nodeTypesAreEqual(actualTypeNode, expectedTypeNode) &&
        astNodesAreEquals(actualTypeNode, expectedTypeNode)
      );
```

(this is definitely not a blocking comment if you prefer the way it is now!)